tests/test-exec-dynamic.cc

```diff
@@ -0,0 +1,150 @@
+  }
+
+  std::vector<char*> av;
+
+};
+
+// define the test suite
+BOOST_FIXTURE_TEST_SUITE(exec_test, ExecTest);
+
+BOOST_AUTO_TEST_CASE( lstm ) {
+  vector<float> results;
+  dynet::ParameterCollection mod;
+  dynet::VanillaLSTMBuilder lstm(2, 3, 10, mod);
+  dynet::LookupParameter lp = mod.add_lookup_parameters(10, {3});
+  dynet::autobatch_flag = 0;
+  dynet::ComputationGraph cg;
+  lstm.new_graph(cg);
+  vector<Expression> losses;
+  for(size_t j = 0; j < 3; ++j) {
+    lstm.start_new_sequence();
+    for(size_t k = 0; k < 3; ++k) {
+      Expression x = dynet::lookup(cg, lp, j*3 + k);
+      lstm.add_input(x);
+    }
+    losses.push_back(squared_norm(lstm.final_h()[1]));
+  }
+  losses.push_back(losses[0] + losses[2]);
+  Expression z = dynet::sum(losses);
+  results.push_back(as_scalar(z.value()));
+
+  for (size_t i = 1; i < results.size(); ++i)
+    BOOST_CHECK_CLOSE(results[0], results[i], 0.0001);
+}
+
+BOOST_AUTO_TEST_CASE( param_after_node ) {
+  auto autobatch_cache = dynet::autobatch_flag;
+  dynet::autobatch_flag = 0;
+  ComputationGraph cg;
+  ParameterCollection model;
+  Parameter param = model.add_parameters({ 1 });
+  
+  Expression loss = zeroes(cg, { 1 });
+  parameter(cg, param);
+  
+  cg.incremental_forward(loss);
+  cg.backward(loss);
+  dynet::autobatch_flag = autobatch_cache;
+}
+
+BOOST_AUTO_TEST_CASE( param_after_node_2 ) {
+  auto autobatch_cache = dynet::autobatch_flag;
+  dynet::autobatch_flag = 0;
+  ComputationGraph cg;
+  ParameterCollection model;
+  LookupParameter param = model.add_lookup_parameters(10, { 1 });
+
+  lookup(cg, param, 1);
+  Expression loss = zeroes(cg, { 1 });
+
+  cg.incremental_forward(loss);
+  cg.backward(loss);
+  dynet::autobatch_flag = autobatch_cache;
+}
+
+BOOST_AUTO_TEST_CASE( multi_lstm ) {
+  vector<vector<float>> results(4);
+  dynet::ParameterCollection mod;
+  dynet::VanillaLSTMBuilder lstm_proto(2, 3, 10, mod);
+  dynet::LookupParameter lp_proto = mod.add_lookup_parameters(10, {3});
+  dynet::autobatch_flag = 0;
+  
+  vector<thread> threads(4);
+  for (size_t t = 0; t < 4; ++t) {
+    threads[t] = thread([&, t]() {
+      dynet::ComputationGraph cg;
+      dynet::VanillaLSTMBuilder lstm(lstm_proto);
+      dynet::LookupParameter lp(lp_proto);
+      lstm.new_graph(cg);
+        
+      vector<Expression> losses;
+      for(size_t j = 0; j < 3; ++j) {
+        lstm.start_new_sequence();
+        for(size_t k = 0; k < 3; ++k) {
+          Expression x = dynet::lookup(cg, lp, j*3 + k);
+          lstm.add_input(x);
+        }
+        losses.push_back(squared_norm(lstm.final_h()[1]));
+      }
+      losses.push_back(losses[0] + losses[2]);
+      Expression z = dynet::sum(losses);
+      results[t].push_back(as_scalar(z.value()));
+    });
+  }
+
+  for (size_t t = 0; t < 4; ++t) { threads[t].join(); }
+  for (size_t t = 0; t < 4; ++t) {
+    for(size_t i = 1; i < results[t].size(); ++i) {
+      BOOST_CHECK_CLOSE(results[t][0], results[t][i], 0.0001);
```

Comment:
This code never runs because results[t].size() is always 1.  Thread safety is never tested (unless the test crashes, which it does).